dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2265,19 +2265,17 @@ void InterpreterSelectQuery::executeOrder(Pipeline & pipeline, SortingInfoPtr so
          */
 
         bool need_finish_sorting = (sorting_info->prefix_order_descr.size() < order_descr.size());
+
+        UInt64 limit_for_merging = (need_finish_sorting ? 0 : limit);
+        executeMergeSorted(pipeline, sorting_info->prefix_order_descr, limit_for_merging);
+
```

Comment:
I'm trying to figure out how this bug works:
With bug: input block streams sorted by A, (1) sort each block by AB, (2) merge streams as if sorted by A, (3) finish sorting the stream by AB as if the input stream is sorted by A (ok) and each block by AB <-- this last assumption is broken by merge.
With fix: input block streams sorted by A, merge streams as if sorted by A, sort each block by AB, finish sorting the stream by AB as if the input stream is sorted by A, and each block by AB -- all assumptions hold.
Is this correct?

As a side note, the names `order_descr` and `sorting_info->prefix_order_desc` make it impossible to tell which one is the input order and which is the required output order.

Reply:
Yes, you are right. 

We can rename `prefix_order_desc` to `read_order_descr` or `input_order_descr`.

Reply:
Maybe something like `input_order` instead of `sorting_info` and `output_order` instead of `order_descr`?